File: cmp-go/go/src/api/ml.go
package api

import (
	"encoding/json"
	"net/http"

	// "reflect"
	// "strconv"

	"io"
	"io/ioutil"

	"goml/core"
	"goml/helper"
	"goml/logger"
	"goml/util"

	tf "github.com/tensorflow/tensorflow/tensorflow/go"
)

var (
	tfModel     *tf.SavedModel
	dataHandler helper.DataSet
	metaHandler helper.DataSet
)

func predict(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	log := logger.GetLogger("api")
	log.Info("Call predict rest api.")

	// get json data from post
	body, err := HttpBodyToMap(r, 1048576)
	if err != nil {
		log.Error("%v", err)
		resposeError(w, err)
		return
	}

	// get data helper
	dataHandler.SetStream(body["sequance"])
	inputData := core.LoadData(dataHandler)

	// apply normalization
	core.LoadData(metaHandler)
	inputData = core.Standardization(inputData, metaHandler.Mean(), metaHandler.Std())

	// convert data to tf tensor
	inputTensor := core.ToTensor(inputData)

	// actually get the prediction from the model
	// pred := core.Predict(tfModel, inputTensor)
	pred := core.Inverse_Standardization(core.Predict(tfModel, inputTensor), metaHandler.Mean(), metaHandler.Std())

	// convert json from data
	doc := util.ToJson(body["label"].(string), pred)
	resposeMessage(w, doc)
	log.Info("Respose Message: %v", string(doc))
}

func resposeError(w http.ResponseWriter, err error) {
	log := logger.GetLogger("api")
	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	w.WriteHeader(422) // unprocessable entity
	if err := json.NewEncoder(w).Encode(err); err != nil {
		log.Error("%v", err)
	}
}

func resposeMessage(w http.ResponseWriter, msg []byte) {
	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	w.WriteHeader(422) // unprocessable entity
	w.Write(msg)
}

func HttpBodyToMap(r *http.Request, s int64) (map[string]interface{}, error) {
	body, err := ioutil.ReadAll(io.LimitReader(r.Body, s))
	if err != nil {
		return nil, err
	}

	if err := r.Body.Close(); err != nil {
		return nil, err
	}

	request := make(map[string]interface{})
	if err := json.Unmarshal(body, &request); err != nil {
		return nil, err
	}
	return request, nil
}

func install(a *App) {
	tfModel = a.Model
	dataHandler = a.Handler
	metaHandler = a.Meta
	a.Route("/ml/estimator", predict, "POST")
}

File: cmp/api/ml.py
from . import engine
from .. import ml
from .. import core
from ..core import workspace, options, logger
from ..data import helper as data_helper
import json


@engine.app.route('/ml/estimator', methods = ['POST'])
def ml_estimator():
    log = logger.get(__name__)

    if not engine.is_json():
        return engine.get_message()
    
    body = engine.get_json()#json 데이터를 받아옴
    if not body:
        return engine.get_message()

    helper = data_helper.get(engine.ws)    
    data = helper.create(body['sequance'])
    if data is None:
        log.error(f"Load data set failed.")
        return engine.get_message()

    # create time window
    window = ml.window_generator(engine.ws, data)
    #log.info(f'input shape: {window.shape()}')
    log.info(f'input shape: {window.shape()}')
    
    
    # predict model
    pred = ml.predict(engine.model, window)
    pred['label'] = body['label']
    pred['predict'] = core.standardization(pred['predict'], engine.ws.mean, engine.ws.std, inverse=True).flatten().tolist()    

    # export json
    pred = json.dumps(pred, indent=4)
    log.info(f'export json: {pred}')
    return engine.Response(pred, mimetype='application/json'), 200



def main(args):
    print('$>', args)

    # initialize engine 
    engine.init(args)

    # initialize logger 
    logger.init(engine.ws)
    log = logger.get(__name__)
    
    helper = data_helper.get(engine.ws)
    if helper is None:
        log.error(f"Not found data helper (data-helper={engine.ws.data_helper})")
        return

    # load model
    engine.load_model(engine.ws, ml)

    app = engine.api()
    app.run()
